fix(links): endpoint links create

TheoPascoli opened the pull request.

skamril committed:
v2.16.8

Merge pull request #2014 from AntaresSimulatorTeam/hotfix/v2.16.8

laurent-laporte-pro committed:
v2.17

Merge pull request #2028 from AntaresSimulatorTeam/release/2.17.0

skamril committed:
v2.17.1

Merge pull request #2057 from AntaresSimulatorTeam/release/v2.17.1

skamril committed:
v2.17.2

Merge pull request #2074 from AntaresSimulatorTeam/release/v2.17.2

skamril committed:
v2.17.3

Merge pull request #2096 from AntaresSimulatorTeam/release/v2.17.3

skamril committed:
v2.17.4

Merge pull request #2104 from AntaresSimulatorTeam/release/v2.17.4

skamril committed:
v2.17.5

Merge pull request #2111 from AntaresSimulatorTeam/release/2.17.5

TheoPascoli committed:
WIP : first commit

TheoPascoli committed:
isort and black

TheoPascoli committed:
add versioning

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add versioning

TheoPascoli committed:
add versioning

TheoPascoli committed:
add versioning

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

MartinBelthle commented:
I see that there are still 68 tests that fail (probably all for the same reason) so I'll let you fix that before reviewing your work :)

TheoPascoli committed:
correct typing issue

TheoPascoli committed:
correct typing issue

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
correct typing issue

TheoPascoli committed:
fix tests

TheoPascoli committed:
fix tests

TheoPascoli committed:
fix tests

TheoPascoli committed:
refactor code

TheoPascoli committed:
refactor code

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -30,8 +31,8 @@ class LinkProperties:
     DISPLAY_COMMENTS: bool = True
-    TRANSMISSION_CAPACITIES: str = "enabled"
-    ASSET_TYPE: str = "ac"
+    TRANSMISSION_CAPACITIES: TransmissionCapacity = "enabled"
```
I think it should be `TransmissionCapacity.ENABLED` and `AssetType.AC`. It shushes mypy issues

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -17,6 +17,7 @@
 from antarest.core.model import JSON
 from antarest.core.utils.utils import assert_this
 from antarest.matrixstore.model import MatrixData
+from antarest.study.storage.rawstudy.model.filesystem.config.links import AssetType, LinkStyle, TransmissionCapacity
```
`LinkStyle` import is unused

MartinBelthle commented on tests/storage/business/test_arealink_manager.py:
```diff
@@ -37,10 +44,11 @@
 from antarest.study.storage.variantstudy.model.dbmodel import VariantStudy
 from antarest.study.storage.variantstudy.model.model import CommandDTO
 from antarest.study.storage.variantstudy.variant_study_service import VariantStudyService
+from tests.conftest_services import study_storage_service_fixture
```
Unused import

MartinBelthle commented on tests/storage/business/test_arealink_manager.py:
```diff
@@ -24,11 +24,18 @@
+from antarest.study.business.link_management import (
+    LinkInfoDTO820,
+    LinkInfoDTOBase,
+    LinkInfoDTOType,
```
Unused import

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/links.py:
```diff
@@ -66,6 +66,42 @@ class TransmissionCapacity(EnumIgnoreCase):
+    DOT_DASH = "dotdash"
+
+
+class FilterOption(EnumIgnoreCase):
```
You should probably use the class `FilteringOptions` inside file /variantstudy/model/command/common.py

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -14,30 +14,152 @@
+
+class LinkInfoFactory:
+    @staticmethod
+    def create_link_info(version: int, **kwargs) -> LinkInfoDTOType:
```
type `**kwargs: t.Any` it will solve mypy issue (same for other warnings in this file)

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -14,30 +14,152 @@
+    @staticmethod
+    def create_parameters(
+        study_version: int, link_creation_info: LinkInfoDTOType
+    ) -> t.Dict[str, t.Union[str, bool, float]]:
```
As filter_synthesis and filter_year_by_year are optional = None you should add None inside the typing Union to fix the mypy issue.
Also change it inside the little following doc

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -52,38 +174,81 @@ class LinkManager:
         command = CreateLink(
             area1=link_creation_info.area1,
             area2=link_creation_info.area2,
+            parameters=LinkInfoFactory.create_parameters(int(study.version), link_info_dto),
```
You could use the variable `study_version` that you've introduced above.

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -52,38 +174,81 @@ class LinkManager:
+            colorg=link_creation_info.colorg,
+            link_width=link_creation_info.link_width,
+            link_style=link_creation_info.link_style,
+            filter_synthesis=link_creation_info.filter_synthesis if study_version >= 820 else None,
```
As you didn't cast link_creation_dto as a LinkInfoDTO820 instance, mypy doesn't know you can ask for this field. 
Little workaround to fix the issue: 
```python
filter_synthesis = None
filter_year_by_year = None
if study_version >= 820:
    link_creation_info = t.cast(LinkInfoDTO820, link_creation_info)
    filter_synthesis = link_creation_info.filter_synthesis
    filter_year_by_year = link_creation_info.filter_year_by_year
```

MartinBelthle reviewed (changes requested):
I'll take a deeper look tomorrow.
I'm not a big fan of having to give every field inside the `create_link_info` method it's a bit heavy in the code IMO (but I don't know how to do differently yet)

MartinBelthle commented:
I believe you should use the pydantic method `model_dump` instead of using the factory.
This method converts a Pydantic object (any class inheriting from BaseModel) into a Dict[str, Any]. 
We use this method everywhere in the code you can take a look to see how it works :)

TheoPascoli committed:
fix mypy

TheoPascoli committed:
fix mypy

TheoPascoli committed:
fix mypy

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
change ui handling

TheoPascoli committed:
black fixing

TheoPascoli committed:
fix errors

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
fix errors

TheoPascoli committed:
fix errors

MartinBelthle committed:
refactor code

TheoPascoli committed:
isort and black

TheoPascoli committed:
refactor classes

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add integration tests

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add integration tests

TheoPascoli committed:
add integration tests

skamril committed:
v2.17.6

Merge pull request #2159 from AntaresSimulatorTeam/release/2.17.6

TheoPascoli committed:
add integration tests

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add integration tests

MartinBelthle commented on tests/variantstudy/model/command/test_create_link.py:
```diff
@@ -52,15 +47,6 @@ def test_validation(self, empty_study: FileStudy, command_context: CommandContex
             }
         ).apply(empty_study)
 
-        with pytest.raises(ValidationError):
```
You should keep this test it's important for people using directly the command

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,173 @@
+            "use-phase-shifter": False,
+        }
+        assert expected == res.json()
+        client.delete(f"/v1/studies/{study_id}/links/{area1_id}/{area2_id}")
```
just do `res =...` and `res.raise_for_status()` to ensure the deletion worked.
The same goes for the other calls inside this test

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,173 @@
+
+        res = client.post(
+            f"/v1/studies/{study_id}/links",
+            json={
```
You could probably store this inside a variable and only add `filter-year-by-year` inside the expected output to avoid duplicated your code.

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,173 @@
+        client.headers = {"Authorization": f"Bearer {user_access_token}"}  # type: ignore
+
+        preparer = PreparerProxy(client, user_access_token)
+        study_id = preparer.create_study("foo", version=880)
```
I'd prefer you use the version 820 or rename the test

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,173 @@
+
+@pytest.mark.unit_test
+class TestLink:
+    @pytest.mark.parametrize("study_type", ["raw", "variant"])
```
It's a good idea to test for `raw` and `variants` but this is not a real fixture. You'll have to do a check at the beginning of the file to really create a variant like this: 

```python
if study_type == "variant":
    study_id = preparer.create_variant(study_id, name="Variant 1")
```
otherwise you're just doing the same test twice.

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,173 @@
+        }
+        assert expected == res.json()
+
+    @pytest.mark.parametrize("study_type", ["raw", "variant"])
```
I think we don't need to parametrize this test, it will save some time

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,173 @@
+        assert 1 == len(res.json())
+        client.delete(f"/v1/studies/{study_id}/links/{area1_id}/{area2_id}")
+
+        # Test create link with same area
```
It could be good to add another test failing because of a wrongly typed enum parameter such as `transmission-capacities`

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,37 +9,80 @@
+
+
+class LinkInfoProperties(BaseModel):
+    hurdles_cost: Optional[bool] = Field(False, alias="hurdles-cost")
```
I believe you could use the method `to_kebab_case` I introduced here to avoid writing aliases at each line.

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,37 +9,80 @@
+
+    @model_validator(mode="before")
+    def validate_colors(cls, values: t.Dict[str, t.Any]) -> t.Dict[str, t.Any]:
+        if type(values) is dict:
```
We often do `if isinstance(values, dict)` instead of checking the type.
I think it's the same thing it's just to keep code consistency

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,37 +9,80 @@
+                "colorg": values.get("colorg"),
+            }
+            for color_name, color_value in colors.items():
+                if color_value is not None and (color_value < 0 or color_value > 255):
```
I don't think `color_value` can be none here but I might be wrong.

Also you could try to remove this validator and just change 
`colorr: Optional[int] = DEFAULT_COLOR`  to 
`colorr: Optional[int] = Field(default=DEFAULT_COLOR, gt=0, lt=255)` it should do the same thing.

TheoPascoli replied:
Indeed it works

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,37 +9,80 @@
+
+
+class LinkInfoProperties820(LinkInfoProperties):
+    filter_synthesis: Optional[str] = Field(None, alias="filter-synthesis")
```
You could also use `to_kebab_case` I think

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,37 +9,80 @@
+    filter_year_by_year: Optional[str] = Field(None, alias="filter-year-by-year")
+
+    @field_validator("filter_synthesis", "filter_year_by_year", mode="before")
+    def validate_individual_filters(cls, value: Optional[str], field: Any) -> Optional[str]:
```
`field` should be typed as an optional str I believe

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,37 +9,80 @@
+    @field_validator("filter_synthesis", "filter_year_by_year", mode="before")
+    def validate_individual_filters(cls, value: Optional[str], field: Any) -> Optional[str]:
+        if value is not None:
+            filter_options = ["hourly", "daily", "weekly", "monthly", "annual"]
```
You could import `FILTER_OPTIONS` to avoid hard-coding this list

TheoPascoli replied:
it triggers a circular import

MartinBelthle replied:
Arf okay so we can keep the code as it is

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -71,12 +114,6 @@ def validate_series(
         new_values = values if isinstance(values, dict) else values.data
         return validate_matrix(v, new_values) if v is not None else v
 
-    @model_validator(mode="after")
```
Same as a comment i wrote already but you should keep this validation for the script team using directly the commands and you could use mode="before" it's always better

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -52,38 +56,73 @@ class LinkManager:
+
+                link_creation_data: t.Dict[str, t.Any] = {"area1": area_id, "area2": link}
+                link_creation_data.update(link_properties)
+                if not with_ui:
```
I've seen with Sylvain we decided that it was better to remove the parameter `with_ui`and always return everything.
So you can adapt this code and the front-end will have to adapt also.

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -52,38 +56,73 @@ class LinkManager:
-    def create_link(self, study: RawStudy, link_creation_info: LinkInfoDTO) -> LinkInfoDTO:
+    def create_link(self, study: RawStudy, link_creation_info: LinkInfoDTOType) -> LinkInfoDTOType:
+        if link_creation_info.area1 == link_creation_info.area2:
+            raise LinkValidationError("Cannot create link on same node")
```
IMO it's better to log `area` instead of `node` and also add the concerned area name

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -52,38 +56,73 @@ class LinkManager:
+        if link_creation_info.area1 == link_creation_info.area2:
+            raise LinkValidationError("Cannot create link on same node")
+
+        study_version = int(study.version)
```
I think you can simplify the code form line 97 to 112 to 
```python
link_data: LinkInfoDTOType
if study_version < 820:
    if isinstance(link_creation_info, LinkInfoDTO820):
        raise LinkValidationError("Cannot specify a filter value for study's version earlier than v8.2")
    link_data = link_creation_info
else:
    link_data = t.cast(LinkInfoDTO820, link_creation_info)
    link_data.filter_synthesis = link_data.filter_synthesis or FilteringOptions.FILTER_SYNTHESIS
    link_data.filter_year_by_year = link_data.filter_year_by_year or FilteringOptions.FILTER_YEAR_BY_YEAR
```

MartinBelthle reviewed (changes requested):
Last round of changes :)
Also, rename your PR with a title fitting the commitlint for when we merge it

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
fix: fix tests

skamril committed:
build(ui): fix issue with build result not working (#2163)

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/master' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation' of https://github.com/AntaresSimulatorTeam/AntaREST into fix/fix-endpoint-links-creation

TheoPascoli committed:
fix: fix tests

TheoPascoli committed:
fix: remove optional typing in models

TheoPascoli committed:
fix: remove optional typing in models

TheoPascoli committed:
fix: remove optional typing in models

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

# Conflicts:
#	antarest/study/business/link_management.py
#	antarest/study/storage/variantstudy/model/command/create_link.py

TheoPascoli committed:
conflict: merge dev

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -51,38 +56,63 @@ class LinkManager:
+                link_creation_data.update(link_properties)
+
+                link_data: LinkInfoDTOType
+                if int(study.version) < 820:
```
Due to this PR https://github.com/AntaresSimulatorTeam/AntaREST/pull/2156 we handle versions differently now. You should do this instead (this goes for every version check you did in this PR):
`StudyVersion.parse(study.version) < STUDY_VERSION_8_2`

TheoPascoli committed:
adapt to new handling of study_versions

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

sylvlecl committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

sylvlecl commented on antarest/study/business/link_management.py:
```diff
@@ -12,31 +12,37 @@
-    color: str
-    width: float
-    style: str
+class LinkInfoDTOBase(AreaInfo, LinkInfoProperties):
```
Class `AreaInfo` seems not used elsewhere, it should be defined here.
Not a big fan of using inheritance that much, areas could also got just in the definition of those 2 classes. But Ok to leave it this way.

TheoPascoli replied:
I had troubles with some tests when removing the area value, especially commands tests. I left it that way

sylvlecl commented on antarest/study/business/link_management.py:
```diff
@@ -12,31 +12,37 @@
-    color: str
-    width: float
-    style: str
+class LinkInfoDTOBase(AreaInfo, LinkInfoProperties):
```
There is an issue with inheriting from `LinkInfoProperties`: we now use kebab case in the API DTOs, whereas we use camel case everywhere for this. We need to keep consistent with this.

TheoPascoli replied:
I overrided the config with camelcase in this class

sylvlecl commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,179 @@
+            "colorg": 112,
+            "colorr": 112,
+            "display-comments": True,
+            "filter-synthesis": "hourly, daily, weekly, monthly, annual",
```
We should use camel case in returned DTOs, for consistency sake with other endpoints

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -9,38 +9,76 @@
+
+class LinkInfoProperties820(LinkInfoProperties):
+    filter_synthesis: t.Optional[str] = None
+    filter_year_by_year: t.Optional[str] = None
```
I think the best approach here, for both fields, would be to use an annotated type with a [custom serializer](https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers) and validation method.
This way, the fields can have their "real" type which is a `List[FilterOption]` while still being serialized and parsed as a `str`.

This will allow to re-use this annotated type in other places if needed.

TheoPascoli replied:
I added an annotated field that performs validation and serialization but the methods are not reusable

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -213,8 +197,11 @@ def _apply(self, study_data: FileStudy) -> CommandOutput:
 
-        self.parameters = self.parameters or {}
-        link_property = CreateLink.generate_link_properties(self.parameters)
+        properties = LinkProperties.model_validate(self.parameters or {})
```
I think here we should parse according to the version, instead of exluding "by hand" the fields later.
If we parse the right version, it will be serialized as expected for this study version.

sylvlecl commented on antarest/study/business/link_management.py:
```diff
@@ -51,38 +57,62 @@ class LinkManager:
+            link_data = LinkInfoDTOBase(**link_info)
+        else:
+            link_data = LinkInfoDTO820(**link_info)
+            if isinstance(link_data, LinkInfoDTO820) and link_data.filter_synthesis is None:
```
we don't need to check the type here, we know it's an instancie of this

sylvlecl commented on antarest/study/business/link_management.py:
```diff
@@ -51,38 +57,62 @@ class LinkManager:
+            link_data = LinkInfoDTO820(**link_info)
+            if isinstance(link_data, LinkInfoDTO820) and link_data.filter_synthesis is None:
+                link_data.filter_synthesis = FilteringOptions.FILTER_SYNTHESIS
+            if isinstance(link_data, LinkInfoDTO820) and link_data.filter_year_by_year is None:
```
Same, type check is redundant

sylvlecl commented on antarest/study/business/link_management.py:
```diff
@@ -51,38 +57,62 @@ class LinkManager:
+            raise LinkValidationError(f"Cannot create link on same area: {link_creation_info.area1}")
+
+        if StudyVersion.parse(study.version) < STUDY_VERSION_8_2 and isinstance(link_creation_info, LinkInfoDTO820):
+            if link_creation_info.filter_synthesis is not None or link_creation_info.filter_year_by_year is not None:
```
I think we should rely on pydantic here instead:
we parse the object depending on the version, pydantic knows what are the valid fields and which ones are not.

sylvlecl commented on antarest/study/business/link_management.py:
```diff
@@ -51,38 +57,62 @@ class LinkManager:
+        else:
+            link_data = LinkInfoDTO820(**link_info)
+            if isinstance(link_data, LinkInfoDTO820) and link_data.filter_synthesis is None:
+                link_data.filter_synthesis = FilteringOptions.FILTER_SYNTHESIS
```
Shouldn't this be the default value in the class instead ?

sylvlecl reviewed (changes requested):
Some change requests, mainly:
- we should use camelCase in the API return, for consistency with other endpoints
- I think we can rely more on pydantic mechanisms for validation / default values of DTOs

TheoPascoli committed:
Refactor to include camelcase to the return object

TheoPascoli committed:
Refactor link DTOs and validation logic

Updated the structure of LinkInfoDTO classes to use Pydantic's ConfigDict for configuration and streamlined validation logic. Removed unnecessary imports and redundant configurations, while ensuring filter fields are checked appropriately for different study versions.

TheoPascoli committed:
Refactor filter values to use a constant list

Use a centralized FILTER_VALUES list for filter_synthesis and filter_year_by_year in LinkInfoProperties820. This change simplifies code maintenance and ensures consistency across filter validations.

TheoPascoli committed:
Refactor filter fields to use enumerations

Replaced filter fields with a comma-separated list of enums for better validation and clarity. This enhances type safety and reduces the risk of invalid filter values being used. The previous string-based validation approach has been removed in favor of utility functions for converting and serializing enum lists.

TheoPascoli committed:
Add a missing newline before LinkInfoProperties820 class

Ensure code readability and consistency by adding a newline before the definition of the `LinkInfoProperties820` class. This change improves the visual separation and organization of the code.

TheoPascoli committed:
Remove unused import and update type annotations

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

# Conflicts:
#	antarest/study/business/link_management.py
#	antarest/study/business/table_mode_management.py

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -26,22 +44,76 @@
+    model_config = ConfigDict(alias_generator=to_kebab_case, populate_by_name=True)
+
+
+comma_separated_enum_list = Annotated[
```
Nice!
I think we should name it following type conventions, in CamelCase, for example `FilterOptionList`

sylvlecl replied:
Also we can move it to the same module where `FilterOption` is defined, because it's likely that we will want to re-use it where `FilterOption` is used (for example in `LinkProperties` etc, but this can be done in a future work, I don't know if it will have more impacts).

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -26,22 +44,76 @@
+comma_separated_enum_list = Annotated[
+    t.List[FilterOption],
+    BeforeValidator(lambda x: validate_filters(x, FilterOption)),
+    PlainSerializer(lambda x: join_with_comma(x)),
```
detail: I think the lambda is not needed:
```python
PlainSerializer(join_with_comma),
```

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -26,22 +44,76 @@
+
+
+def validate_filters(
+    filter_value: Union[List[FilterOption], str], enum_cls: t.Type[FilterOption]
```
The function only works for FitlerOption class, so I don't think the second argument is useful ?
It would be if we wanted to use this function for other enum types, but here we have hard coded messages, so I don't think we will do it.

sylvlecl replied:
Note:
I found out there are similar methods in `antarest/study/storage/rawstudy/model/filesystem/config/field_validators.py`, we should try to clean this code duplication at some point (we can create a specific task for this if not in that PR)

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -26,22 +44,76 @@
+
+comma_separated_enum_list = Annotated[
+    t.List[FilterOption],
+    BeforeValidator(lambda x: validate_filters(x, FilterOption)),
```
detail: if we remove the second argument, lambda is not needed anymore

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -26,22 +44,76 @@
 from antarest.study.storage.variantstudy.model.model import CommandDTO
 
+DEFAULT_COLOR = 112
+FILTER_VALUES: t.List[FilterOption] = [
```
maybe `DEFAULT_FILTERS` for homogeneity with the above var? indeed it's only used as the default value

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -98,51 +170,6 @@ def _create_link_in_config(self, area_from: str, area_to: str, study_data: FileS
         )
 
-    @staticmethod
-    def generate_link_properties(parameters: JSON) -> JSON:
```
Great to have removed this :)

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -211,12 +229,18 @@ def _apply(self, study_data: FileStudy, listener: Optional[ICommandListener] = N
+        else:
+            properties = LinkInfoProperties.model_validate(self.parameters or {})
+
+        link_property = properties.model_dump(mode="json", exclude={"area1", "area2"}, by_alias=True, exclude_none=True)
```
The excludes are not needed I think, the properties classes don't have information about areas

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -211,12 +229,18 @@ def _apply(self, study_data: FileStudy, listener: Optional[ICommandListener] = N
+
+        properties: LinkInfoProperties
+        if StudyVersion.parse(version) >= STUDY_VERSION_8_2:
+            properties = LinkInfoProperties820.model_validate(self.parameters or {})
```
Taking a step back about the architecture of our classes:
our logic is wrong here, the parameters were created by dumping a `LinkInfoDTO`, but we read it as a `LinkInfoProperties`. There is a kind of implicit conversion.

Instead, we should read it again as a DTO, and have an explicit conversion function to get a "properties" class that will be targetted at being serialized in the INI file (even if it only does a modeul_dump/model_validate to achieve that conversion).

sylvlecl commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -211,12 +229,18 @@ def _apply(self, study_data: FileStudy, listener: Optional[ICommandListener] = N
-        self.parameters = self.parameters or {}
-        link_property = CreateLink.generate_link_properties(self.parameters)
-
         study_data.tree.save(link_property, ["input", "links", area_from, "properties", area_to])
```
Again, taking a step back:
here we are serializing basically the content of a `LinkInfoProperties` into an ini file.
This seems like a duplication of the role of the `LinkProperties` class in `antarest/study/storage/rawstudy/model/filesystem/config/links.py`.

It looks like maybe the new classes you have created should go there and replace the old one.
OR, we should have a conversion function from the new classes to this one, and use it for serialization.

**I am not sure of the best choice right now, we need to think about this from a general point of view, so that we use common principles everywhere.** 
I think we should use this work as an opportunity to clarify and agree about this design of classes with all the team.

sylvlecl reviewed (changes requested):
Sorry, further comments and thoughts, this work makes me realize we should clarify the architecture of our classes to have a common understanding (*Properties, *DTO, their roles, ...).

skamril committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

skamril committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

skamril commented on antarest/study/web/study_data_blueprint.py:
```diff
@@ -151,15 +151,14 @@ def get_areas(
         uuid: str,
-        with_ui: bool = False,
         current_user: JWTUser = Depends(auth.get_current_user),
     ) -> t.Any:
```
`t.List[LinkInfoDTO]` instead?

skamril commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -26,22 +44,76 @@
+    hurdles_cost: bool = False
+    loop_flow: bool = False
+    use_phase_shifter: bool = False
+    transmission_capacities: TransmissionCapacity = TransmissionCapacity.ENABLED
```
`transmission_capacity` in singular (there is an another enum named TransmissionCapacities)

MartinBelthle replied:
The issue is the field is written `transmission-capacities` inside the ini file :/ 
It's a shame that there's another field named the same way inside the generaldata.ini file ...
